source/tutorials/nix-language.md

```diff
@@ -0,0 +1,325 @@
+
+Example:
+
+    nix-repl> let
+              f = x: x + 1;
+              in
+              f 1
+    2
+
+Arguments can be chained.
+
+    x: y: x + y
+
+The above function takes one argument and returns a function `y: x + y` with `x` set to passed value. 
+
+Example:
+
+    nix-repl> let
+              f = x: y: x + y;
+              in
+              f 1
+    «lambda @ (string):2:8»
+
+The `lambda` indicates the resulting value is a function.
+
+Applying that to another argument yields the inner body `x + y`, which can now be fully evaluated.
+
+    nix-repl> let
+              f = x: y: x + y;
+              in
+              f 1 2
+    3
+
+### Keyword arguments
+
+Nix functions can explicitly take an attribute set as argument.
+
+    {a, b}: a + b
+
+This is equivalent to
+
+    x: x.a + a.b
+
+The argument defines the exact attributes that have to be in that set.
+Leaving out or passing additional attributes is an error.
+
+Example:
+
+    nix-repl> let
+              f = {a, b}: a + b
+              in
+              f { a = 1; b = 2; }
+    3
+
+### Default attributes
+
+Also known as “default arguments”.
+
+Arguments can have default values for attributes, denoted with a question mark “?”.
+
+    {a, b ? 0}: a + b
+
+Attributes in the argument are not required if they have a default value.
+
+Example:
+
+    nix-repl> let
+              f = {a, b ? 0}: a + b
+              in
+              f { a = 1; }
+    1
+
+Example:
+
+    nix-repl> let
+              f = {a ? 0, b ? 0}: a + b
+              in
+              f { } # empty attribute set
+    0
+
+### Additional attributes
+
+You can allow additional attributes with an ellipsis (`...`):
+
+    {a, b, ...}: a + b
+
+Example:
+
+    nix-repl> let
+              f = {a, b, ...}: a + b
+              in
+              f { a = 1; b = 2; c = 3; }
+    3
+
+### Named keyword arguments
+
+Also known as “@ syntax” or “‘at’ syntax”.
+
+    {a, b, ...}@args: a + b + args.x
+
+where additional attributes are subsumed under a name.
+
+Example:
+
+    nix-repl> let
+              f = {a, b, ...}@args: a + b + args.x
+              in
+              f { a = 1; b = 2; c = 3; x = 5 }
+    8
+
+This can be useful if this remaining attribute set needs to be processed as a whole.
+
+[^1]: For example, the built-in function `throw` causes evaluation to stop. However, the following works fine, because `throw` is never evaluated:
+     nix-repl> let lazy = { a = "success"; b = builtins.throw "error"; }; in lazy.a
+"success"
+
+## Summary
+
+As a programming language, Nix is
+
+- *declarative*
+
+  It has no notion of executing sequential steps.
+  Dependencies between operations are established only through data.
+  Everything in Nix is an expression that results in a single value.
+  Every Nix file (`.nix`) contains a single expression.
+
+- *purely functional*
+
+  Pure means: Nix does not change the value of declarations during computation – there are no variables, only names for immutable values.
```

Comment:
> there are no variables

Sounds very weird, because there are, but they aren't variable. Maybe reword to "There are no mutable variables"

Reply:
"variables that are not variable" is an oxymoron. Therefore there are no variables in the Nix language.

Reply:
Even if variables never change during execution, they are still called variables in the sense that their value can vary between executions. See https://en.wikipedia.org/wiki/Variable_(mathematics) or https://www.merriam-webster.com/dictionary/variable

Reply:
I think a focus on whether Nix does or does not have "actual variables" is a confusing distraction when describing what pure functional programming means.

Colloquial use of the word "variable", at least among English speakers, applies in an intuitive way to functional programming. Nix has these in function arguments and let bindings, both of which allow a program to cause one name to take on different values. This is the core concept behind what a "variable" is.

See also https://nixos.org/manual/nix/stable/expressions/language-constructs.html#let-expressions which says "A let-expression allows you to define local variables for an expression."

See also https://nixos.org/guides/nix-pills/basics-of-language.html#idm140737320512656 which shows you how to use a let expression to shadow variables, effectively making one name refer to multiple values by introducing a new scope.

There is also https://en.wikipedia.org/wiki/Variable_(computer_science) which says "In pure functional programming, variables are bound to expressions and keep a single value during their entire lifetime due to the requirements of referential transparency."

I would suggest that the description of "pure" avoid the problem entirely by not focusing on a "variables that are not variable" oxymoron. Instead, focus on the idea that functions always return the same when passed the same arguments, and that computed values cannot be modified.  Or something like that...

Reply:
I will remove this section entirely. I had been influenced by the manual, [which says](https://web.archive.org/web/20220716000213/https://nixos.org/manual/nix/stable/expressions/expression-language.html)

> for instance, there is no variable assignment

The manual should be the reference on what the Nix language is and does, so I moved my attempt to describe it more clearly there: https://github.com/NixOS/nix/pull/6907

The [NixOS Wiki on Nix language replicates this](https://web.archive.org/web/20220812152834/https://nixos.wiki/wiki/Overview_of_the_Nix_Language), and that redundancy should IMHO go away, and instead we should always refer to the manual. Same for this tutorial.